### PR TITLE
Analyze and report 'pub downgrade' issues.

### PR DESCRIPTION
**Pull request by isoos**
- #1356

### CODE REVIEW

**Comment by jonasfj** [comment]
I propose granted 20 points in the "Pass static analysis" section, as follows:
image

@mit-mit Any thoughts on giving 20 points when passing static analysis after `dart pub downgrade`. That's trying with a set of oldest versions that:
 * Is allowed by constraints in `pubspec.yaml`, and,
 * Is mutually compatible.

This effectively passing `dart pub downgrade && dart analyze`, do notice that downgrade will try to get an oldest set of versions, there may be multiple sets of oldest versions, and it may not be possible get oldest versions of some packages (due to incompatibilities, etc).

In an ideal world, we could prove that a package is compatible with the lower-bounds of its dependencies. This is an under-approximation such a test. But it's likely to catch a lot!

**Comment by jonasfj** [comment]
note: package authors can always fix this issue by running `dart pub upgrade --tighten` and republish a new version.

**Comment by isoos** [comment]
Note: I've left the subsection inside the up-to-date dependencies section and set 20 points for it.

File: test/goldens/end2end/dnd-2.0.1.json_report.md [hunk 1]
```diff
@@ -127,3 +127,8 @@ To reproduce run `dart pub outdated --no-dev-dependencies --up-to-date --no-depe
 
 
 ### [*] 10/10 points: Package supports latest stable Dart and Flutter SDKs
+
+
+### [*] 20/20 points: Dependency constraint lower bounds are not breaking
+
+`pub downgrade` does not expose any static analysis error.
```
Comment (sigurdm) [inline test/goldens/end2end/dnd-2.0.1.json_report.md:134]: Idea: Should we expose the versions that downgrade ends up with?

File: test/goldens/end2end/nsd_android-1.2.2.json_report.md [hunk 2]
```diff
@@ -68,3 +68,20 @@ So, because nsd_android depends on flutter from sdk, version solving failed.
 ```
 
 ### [*] 10/10 points: Package supports latest stable Dart and Flutter SDKs
+
+
+### [x] 0/20 points: Dependency constraint lower bounds are not breaking
```
[the remark this answers is not in the document]
Comment (isoos) [reply test/goldens/end2end/nsd_android-1.2.2.json_report.md]: We don't really lose much analysis time with the attempt, worst case we have the same error message as here. Unless this gets really long, we should just keep repeating it.

**Comment by isoos** [comment]
Updated suggestion to "You may run `dart pub upgrade --tighten` to update your dependency constraints, or visit http://dart.dev/go/downgrade-testing for further help." PTAL.

**Review by sigurdm — APPROVED** [review]

**Pull request merged**